tiledb/sm/misc/resource_pool.h

```diff
@@ -139,6 +144,94 @@ class ResourcePool {
   friend class ResourceHandle<T, tiledb::sm::ResourcePool>;
 };
 
+template <class T>
+class BlockingResourcePool {
+ public:
+  using resource_handle = ResourceHandle<T, tiledb::sm::BlockingResourcePool>;
+
+  /** Constructor. */
+  BlockingResourcePool(unsigned int n)
+      : resources_(n)
+      , unused_(n)
+      , unused_idx_(n - 1)
+      , num_blocked_threads_(0) {
+    for (unsigned int i = 0; i < n; i++)
+      unused_[i] = i;
+  }
+
+  /** Take a resource from the pool. */
+  resource_handle take() {
+    if (resources_exhausted()) {
+      // block until available again
+      std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
+      num_blocked_threads_++;
+      exhaustion_cv_.wait(e_lck, [this]() { return !resources_exhausted(); });
+      num_blocked_threads_--;
+    }
+    return add_resource();
+  }
+
+ private:
+  /** Release a resource from the pool. */
+  void release(unsigned int n) {
+    remove_resource(n);
+    if (blocked_threads())
+      exhaustion_cv_.notify_one();
+  }
+
+  /** Access a resource from the internal vector. */
+  T& at(unsigned int n) {
+    return resources_.at(n);
+  }
+
+  /** Check if resource pool is full */
+  inline bool resources_exhausted() {
+    return (unused_idx_ == -1);
+  }
+
+  /** Check if there is any blocked thread */
+  inline bool blocked_threads() {
+    std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
```

Comment:
Obtaining a lock doesn't have any effect here, since there's only one datum whose value is referenced.

As a rule, you need a lock when you need to do two or more things simultaneously, such as maintaining a class invariant.

Reply:
removed lock